src/baltree.h

```diff
@@ -45,26 +60,76 @@
 #ifndef BALANCED_TREE_INIT    // non-generic part shared by all versions
 #define BALANCED_TREE_INIT
 
+// Scapegoat trees are height-balanced trees. With respect to a
+// parameter alpha, a binary tree is height-balanced, if
+//
+//   height(tree) <= log_1/alpha(size(tree))
+//
+// where size(tree) is the number of nodes that a tree contains.
+//
+// A binary tree is called weight-balanced iff for any subtree `tree`,
+//
+//   size(left) <= alpha * size(tree) and
+//   size(right) <= alpha * size(tree)
+//
+// where `left` and `right` are the left and right subtrees of `tree`.
+//
+// Any weight-balanced tree is also height-balanced, but not all
+// height-balanced trees are also weight-balanced. However, by
+// contraposition, any tree that is not height-balanced cannot be
+// weight-balanced, either.
+//
+// Furthermore, for any tree that is not height-balanced, any node
+// at maximum depth will have an ancestor that is not weight-balanced.
+// Thus, if we violate the height balance property through the insertion
+// of a new node, we can simply check that node's ancestors to look
+// for a node that is not weight-balanced and rebalance the subtree
+// rooted at that node.
+//
+// Rebalancing operations have time complexity linear in the number
+// of nodes affected, but their amortized cost is O(log n) per node
+// for a tree with n nodes, as most insertions either do not
+// require rebalancing or only the rebalancing of a small number of
+// nodes.
+//
+// Overall, this ensures that height(tree) = O(log(size(tree)) and
+// that insertion, deletion, and lookup operations can be done in
+// amortized O(log(size(tree)) time. While individual insertion or
+// deletion operations (though not lookup operations) can take
+// O(size(tree)) time, the amortized time complexity remains
+// logarithmic.
+//
+// See [Galperin and Rivest 1993] for further details and proofs
+// of the above claims.
+
+// The value of MaxTreeDepth assumes that alpha <= 1/sqrt(2). Larger
+// values of alpha do not lead to well-balanced trees.
 enum { MaxTreeDepth = 2 * (sizeof(UInt) * 8) };
 
-static int height_to_size_init = 0;
+static int min_nodes_for_height_init = 0;
 // For scapegoat trees with balance factor alpha:
-// height_to_size: d -> (1/alpha) ^ d
-static Int height_to_size[MaxTreeDepth];
+// min_nodes_for_height: d -> (1/alpha) ^ d
+//
+// This array contains the minimum number of nodes that
+// a binary tree of a given height can contain while still
+// being height-balanced. Any tree with fewer nodes is too
+// sparse to be height balanced.
+static Int min_nodes_for_height[MaxTreeDepth];
 
 // alpha = alpha_lo / alpha_hi
+// We use integers to allow for more efficient arithmetic.
 static const Int alpha_hi = 3;
 static const Int alpha_lo = 2;
```

Comment:
if efficiency is a concern, perhaps `int` instead of `Int` would be better? Not that I checked the actual machine code, but that's the rule of thumb, isn't it?

(I guess in the context of GAP, I'd actually make this an `enum { ALPHA_NUM = 3, ALPHA_DEN = 2 };` but this code is outside the scope of this PR, so this does not have to be touched by this PR. I.e.: feel free to ignore, this is just me rambling on ;-)

Reply:
`Int` vs `int` doesn't matter here, as they're constants, but I have made them enums to be consistent with the rest of GAP.